actionpack/lib/action_dispatch/middleware/cookies.rb

```diff
@@ -625,39 +673,42 @@ def commit(options)
-    class UpgradeLegacyEncryptedCookieJar < EncryptedCookieJar #:nodoc:
-      include VerifyAndUpgradeLegacySignedMessage
-    end
+        def try_legacy_verifier(name, encrypted_message)
+          deserialize(name, @legacy_verifier.verify(encrypted_message)).tap do |deserialize|
+            self[name] = deserialize
+          end
 
-    # UpgradeLegacyHmacAesCbcCookieJar is used by ActionDispatch::Session::CookieStore
-    # to upgrade cookies encrypted with AES-256-CBC with HMAC to AES-256-GCM
-    class UpgradeLegacyHmacAesCbcCookieJar < EncryptedCookieJar
-      def initialize(parent_jar)
-        super
+        rescue ActiveSupport::MessageVerifier::InvalidSignature
+          nil
+        end
 
-        secret = key_generator.generate_key(request.encrypted_cookie_salt || "")[0, ActiveSupport::MessageEncryptor.key_len]
-        sign_secret = key_generator.generate_key(request.encrypted_signed_cookie_salt || "")
+        def expand_outbound_rotation(outbound)
+          raise ArgumentError, "Must provide a :cipher value" unless outbound.key?(:cipher)
+          raise ArgumentError, "Must provide either a :raw_key or :secret value" unless outbound.key?(:raw_key) || outbound.key?(:secret)
+          raise ArgumentError, "Must provide a :salt value when using :secret" if outbound.key?(:secret) && !outbound.key?(:salt)
 
-        @legacy_encryptor = ActiveSupport::MessageEncryptor.new(secret, sign_secret, cipher: "aes-256-cbc", digest: digest, serializer: ActiveSupport::MessageEncryptor::NullSerializer)
-      end
+          has_signed_key = outbound.key?(:signed_salt) || outbound.key?(:raw_signed_key)
 
-      def decrypt_and_verify_legacy_encrypted_message(name, signed_message)
-        deserialize(name, @legacy_encryptor.decrypt_and_verify(signed_message)).tap do |value|
-          self[name] = { value: value }
-        end
-      rescue ActiveSupport::MessageVerifier::InvalidSignature, ActiveSupport::MessageEncryptor::InvalidMessage
-        nil
-      end
+          if outbound[:cipher] =~ AES_CBC_CIPHER_REGEX && !has_signed_key
+            raise ArgumentError, "Must provide :signed_salt or :raw_signed_key when using AES-CBC"
+          end
 
-      private
-        def parse(name, signed_message)
-          super || decrypt_and_verify_legacy_encrypted_message(name, signed_message)
+          key_len = ActiveSupport::MessageEncryptor.key_len(outbound[:cipher])
+
+          key = outbound[:raw_key]
+          key ||= ActiveSupport::KeyGenerator.new(outbound[:secret], iterations: 1000).generate_key(outbound[:salt], key_len)
+
+          options = outbound.slice(:cipher, :digest).update(serializer: SERIALIZER)
+
+          if has_signed_key
+            sign_key = outbound[:raw_signed_key]
+            sign_key ||= ActiveSupport::KeyGenerator.new(outbound[:secret], iterations: 1000).generate_key(outbound[:salt])
+
+            return key, sign_key, options
+          end
+
+          return key, options
```

Comment:
Also think this is doing too much. How about:

```ruby
Array.wrap(encrypted_cookie_rotate_out).each do |rotation|
  config = rotation.slice(:cipher, :digest).merge(serializer: serializer)

  @encryptor.rotate(extract_signing_key(rotation), extract_key(rotation), config)
end

if rotate_out_legacy_aes_cbc_hmac_cookies?
  config = { cipher: "aes-256-cbc", digest: digest }
  
  @encryptor.rotate(
    extract_signing_key(config.merge(key_generator: key_generator, signed_salt: encrypted_signed_cookie_salt),
    extract_key(config.merge(key_generator: key_generator, salt: encrypted_cookie_salt)),
    config
  )
end

def extract_key(cipher:, raw_key: nil, key_generator: nil, secret: nil, salt: nil, **)
  if raw_key
    raw_key
  else
    raise ArgumentError, "Missing salt to generate encryption key from" unless salt

    key_generator ||= ActiveSupport::KeyGenerator.new(secret, iterations: 1000)
    key_generator.generate_key(salt, ActiveSupport::MessageEncryptor.key_len(cipher))
  end
end

def extract_signing_key(cipher:, raw_signed_key: nil, key_generator: nil, secret: nil, signed_salt: nil, **)
  if cipher =~ AES_CBC_CIPHER_REGEX && !(raw_signed_key || signed_salt)
    raise ArgumentError, "Missing signed_salt or raw_signed_key when using AES-CBC"
  end

  if raw_signed_key
    raw_signed_key
  else
    raise ArgumentError, "Missing signed_salt to generate sign key from" unless signed_salt

    key_generator ||= ActiveSupport::KeyGenerator.new(secret, iterations: 1000)
    key_generator.generate_key(signed_salt)
  end
end
```

Reply:
Refactored this a bit more and added a couple more test cases to the cookie middleware for rotating encrypted cookies.